tensorflow_probability/python/distributions/empirical.py

```diff
@@ -0,0 +1,387 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+"""The Empirical distribution class."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+from tensorflow.python.ops import gen_array_ops
+from tensorflow_probability.python.distributions import distribution
+from tensorflow_probability.python.internal import distribution_util as util
+from tensorflow_probability.python.internal import reparameterization
+
+__all__ = [
+    "Empirical",
+    "VectorEmpirical"
+]
+
+
+def _broadcast_event_and_samples(event, samples, event_ndims):
+  """Broadcasts the event or samples."""
+  samples_shape = tf.shape(samples)[:-event_ndims-1]
+  event *= tf.ones(samples_shape, dtype=event.dtype)
+  event = tf.expand_dims(event, axis=-event_ndims-1)
+  samples *= tf.ones_like(event, dtype=samples.dtype)
+
+  return event, samples
+
+
+class _BaseEmpirical(distribution.Distribution):
+  """Base class for Empirical distribution."""
+
+  def __init__(self,
+               samples,
+               event_ndims=0,
+               validate_args=False,
+               allow_nan_stats=True,
+               name="Empirical"):
+    """Initialize `_BaseEmpirical` distributions.
+
+    Args:
+      samples: Numeric `Tensor`. Samples or batches of samples on which the
+        distribution is based.
+      event_ndims: Python `int32`, default `0`. number of event dimensions.
+        When `0` this distribution is for `Empirical`. When `1` this
+        distribution is for `VectorEmpirical`.
+      validate_args: Python `bool`, default `False`. When `True` distribution
+        parameters are checked for validity despite possibly degrading runtime
+        performance. When `False` invalid inputs may silently render incorrect
+        outputs.
+      allow_nan_stats: Python `bool`, default `True`. When `True`, statistics
+        (e.g., mean, mode, variance) use the value "`NaN`" to indicate the
+        result is undefined. When `False`, an exception is raised if one or
+        more of the statistic's batch members are undefined.
+      name: Python `str` name prefixed to Ops created by this class.
+
+    Raises:
+      ValueError: if `samples` is a scalar.
+    """
+
+    parameters = locals()
+    with tf.name_scope(name, values=[samples]):
+      self._samples = tf.convert_to_tensor(samples, name="samples")
+      self._event_ndims = event_ndims
+      with tf.control_dependencies([tf.assert_rank_at_least(self._samples, 1)]):
```

Comment:
should this be `with tf.control_dependencies([tf.assert_rank_at_least(self._samples, event_ndims + 1)]):`

Reply:
Good point. I was thinking about some fault-tolerance for samples. For example, VectorEmpirical(samples=[0, 1]) would be broadcast to VectorEmpirical(samples=[[0, 1]]). Do you think it is necessary?

Reply:
I think it's more helpful to have `VectorEmpirical(samples=[0, 1])` cause an error. If you really want a single vector, you could use `Deterministic` or you could put the extra brackets around `samples`.

Reply:
I agreed. I modified the code and docstring to address it.